### DISCUSSION TITLE
Yanking and pasting to system clipboard not working under Wayland

### DISCUSSION THREAD

**Discussion by Provisional9448** [helix-editor/helix#8755] [Troubleshooting] [answered]
I'm trying to copy/paste between Helix and other apps, but I get an error `clipboard provider wl-copy failed` when using `space+y` or `:clipboard-yank`. Copying and pasting  works fine if I run `sudo hx`.

I think it might have something to do with `wl-copy` and `wl-paste` and how my environment variables are set up. I did check that `wl-copy` and `wl-paste` are in my `$PATH`. `which wl-copy` results in `/usr/local/bin/wl-copy`. 

When I run `wl-copy hello world!`, I get an error `Missing a shell`. And when I run `sudo wl-copy hello world!`, I get an error:
```
error: XDG_RUNTIME_DIR is invalid or not set in the environment.
Failed to connect to a Wayland server
```

Is this just a Wayland issue or is there something I can do to fix this?

Thanks!

Edit: I just found out that `hx --health clipboard` returns `System clipboard provider: wl-paste+wl-copy` but running it as `sudo` I get `System clipboard provider: xsel`. 

Why is the `sudo` system clipboard using `xsel` when I'm on Wayland?

**Comment by pascalkuthe** [comment]
It's hard to tell what is wrong here but this is not a helix specific issue. The clipboard providers are clearly not working so something with your setup is wrong. 

If xsel is working for you it actually sounds like you are using x-org and not wayland tough.

**Reply by Provisional9448** [reply]
I am definitely using Wayland, because in the `About` section in settings it says Wayland. If I inspect windows using `alt+F2` and type `lg` to open Gnome Looking Glass, I can see my windows are in `MetaWindowWayland`. 

I did find someone with a [similar issue with Neovim posted on the `wl-clipboard` repo.](https://github.com/bugaevc/wl-clipboard/issues/96#issuecomment-696117547) 

> This is probably because `sudo` messes with the environment, so you end up with `XDG_RUNTIME_DIR` unset. It should fail like this:
> 
> ```
> $ sudo wl-paste 
> error: XDG_RUNTIME_DIR not set in the environment.
> Failed to connect to a Wayland server
> ```
> 
> Try using `sudo -E` to keep the environment.


It seems like this may be an intended functionality of how copy and paste works in Wayland. As of right now, I'll temporarily run `hx` as `sudo hx` so the clipboard will work. 

---

Edit: For some reason `xsel` is working while `wl-clipboard` is not. I have confirmed that I am, in fact, using Wayland. Is there a way to set `xsel` as the system clipboard provider for `hx`?

**Reply by Provisional9448** [reply, marked as answer]
Okay, I have got it working now. The reason why `wl-copy` and `wl-paste `were not working is because I had two versions on my system, likely from a prior build. 

If anyone comes across the same problem, have them verify that `wl-copy` and `wl-paste` are in both the `/usr/local/bin` and `/usr/bin` directories. Remove the ones from `/usr/local/bin` in order to default to the` /usr/bin` version from the distro's repo. 

This is where I found the solution: 
https://github.com/bugaevc/wl-clipboard/issues/181#issuecomment-1608386370

**Reply by dpvpro** [reply]
I have same issue and resolve it by removing X related packages

```
yay -Rnsu xclip xsel
```

```
user@host ~> sudo hx --health clipboard
System clipboard provider: termcode
```
